### ISSUE TITLE
Compiler should report diagnostics for more def-assign locations in Local-Function scenarios.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#17769]
Right now for the following code:

```c#
        F();  // Error about def-assignment is only reported here.
        void F()
        {
            Action f = () =>
            {
                Dictionary<int, int> dict = null;
                dict?.TryGetValue(0, out var x);
                Console.WriteLine(x);
            };
        }
```

Both for IDE purposes and for user clarity we should have something more akin to:

```c#
        F();  // Expected error here that F can't be called because there are def-assign issues.
        void F()
        {
            Action f = () =>
            {
                Dictionary<int, int> dict = null;
                dict?.TryGetValue(0, out var x);
                Console.WriteLine(x); // error here that 'x' would not be definitely assigned
            };
        }
```

**Comment by CyrusNajmabadi** [comment]
Tagging @agocke

**Comment by CyrusNajmabadi** [comment]
With this work the IDE would be able to tell that inlining certain variables would cause issues.  See: 
https://github.com/dotnet/roslyn/issues/17743
